Node bot

vitsalis opened the pull request.

vitsalis committed:
[NODE_BOT] Add a basic node bot

vitsalis committed:
Make game a post request

dionyziz commented on bots/node/package.json:
```diff
@@ -0,0 +1,22 @@
+    "test": "mocha"
+  },
+  "author": "Vitalis Salis",
+  "license": "ISC"
```
Please release your code under the MIT license.

vitsalis committed:
Merge branch 'GET_on_bot_initialize' into node_bot

dionyziz commented:
Your tests should be automated and included in travis.

vitsalis committed:
[NODE_BOT] Use the MIT license instead of the ISC one

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    var round = req.body.round;
+    var map = JSON.parse( req.body.map );
```
This can throw if `req.body.map` isn't well formed and will take down the server with it. Needs a `try{} catch {}` and an appropriate response code to the user (probably 400)

petrosagg replied:
An other note here is that, because you're using the `bodyParser` middleware, if the POST request is sent with the correct Content-Type header (i.e `application/json`) then `req.body.map` will already be an object because `bodyParser` parses it for you. I think the best approach to this is testing whether the `Content-Type` header is `application/json` and if not return a 400 status code without further processing.

dionyziz replied:
Requires #359; will merge for now, we will work on this on a separate PR.

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+        version: '0.1.0',
+        username: 'sample_username'
+    } ) );
+    res.status( 200 ).end();
```
You don't need to specify `status` or call `end()` when you've used the `send()` method

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+
+app.post( '/game', function( req, res ) {
+    res.send( JSON.stringify( {} ) );
+    res.status( 200 ).end();
```
Same here

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+            ];
+
+            for ( var j = 0; j < offsets.length; ++j ) {
+                var newX = parseInt( x ) + offsets[ j ][ 0 ], newY = parseInt( y ) + offsets[ j ][ 1 ];
```
It's a good practice to pass a second argument to `parseInt` which is the base of the number you're trying to parse. This is to prevent inputs like '0312' to be treated as octal numbers. i.e `parseInt(x, 10)` instead of `parseInt(x)`

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+    res.send( JSON.stringify( {
+        'intent': intent
+    } ) );
+    res.status( 200 ).end();
```
Same here for `status` and `end`

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+
+app.listen( 8000 );
+
+app.get( '/bot', function( req, res ) {
```
I'd suggest that API calls be prefixed with an API version like `/v1/bot` so that it's clear when you change it and/or make backwards incompatible changes.

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+    res.status( 200 ).end();
+} );
+
+app.post( '/round', function( req, res ) {
```
This function uses user input without any sort of validation. Any trivially malformed input will cause a `TypeError` exception when trying to read properties from undefined values which will again take the whole server down with it.

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+                } );
+            }
+            else {
+                if ( !moves.length || Math.random() * 2 < 1 ) {
```
Why not `Math.random() < 0.5` ?

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,113 @@
+app.listen( 8000 );
+
+app.get( '/bot', function( req, res ) {
+    res.send( JSON.stringify( {
```
You can use the `res.json()` method and `express` will handle the serialisation for you.

petrosagg replied:
Same for all other methods in this file

petrosagg replied:
Plus, it will correctly set the response header to `application/json`

vitsalis committed:
[NODE_BOT] Regression test and fix bug

The bot was attacking dead enemy creatures.
Beating a dead horse is not going to get you anywhere.

petrosagg commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,311 @@
+        //console.log( JSON.parse( JSON.stringify( map ) ) );
+        agent
+            .post( '/round' )
+            .send( 'map=' + JSON.stringify( map ) +'&myid=1&W=10&H=10&round=1&gameid=1' )
```
This `JSON.stringify( map )` should probably be wrapped in `encodeURIComponent()` to avoid weird errors.

petrosagg commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,311 @@
+        agent
+            .post( '/round' )
+            .send( 'map=' + JSON.stringify( map ) +'&myid=1&W=10&H=10&round=1&gameid=1' )
+            .expect( function( res ) {
```
Add these two before the expect call?

```
  .set('Accept', 'application/json')
  .expect('Content-Type', /json/)
```

vitsalis committed:
[NODE_BOT] Use res.json method instead of res.send for json

vitsalis committed:
[NODE_BOT] Remove status and end calls

vitsalis committed:
[NODE_BOT] Specify the base on parseInt

vitsalis committed:
[NODE_BOT] Use a smarter method for random choice

Math.random() < 0.5 instead of Math.random() * 2 < 1

vitsalis committed:
[NODE_BOT] Make expected content-type json

vitsalis committed:
[NODE_BOT] Wrap JSON.stringify(map) inside encodeURIComponent()

dionyziz commented on bots/node/package.json:
```diff
@@ -0,0 +1,22 @@
+    "express": "^4.8.0",
+    "multer": "^0.1.3",
+    "mocha": "^1.4.14",
+    "request": "^2.39.0"
```
Please include `supertest`.

dionyziz commented:
Node best practices recommend a README file. Please include a simple one. A one-line description is OK.

vitsalis committed:
[NODE_BOT] Include supertest on package.json

dionyziz commented:
mocha tests are passing for this PR, ready to merge test-wise.

vitsalis committed:
[NODE_BOT] Regression test that bot responds with 400 HTTP code when invalid json is sent

vitsalis committed:
[NODE_BOT] Add a README file

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,116 @@
+        var map = JSON.parse( req.body.map );
+    }
+    catch ( e ) {
+        res.status( 400 ).end();
```
You can use `res.send(400)`

dionyziz replied:
For reference, `res.send(400)` is deprecated, so the patch looks good. Will review further tonight.

pkakelas commented:
lgtm

dionyziz commented:
Test automation postponed for a separate PR as of #361; will merge this for now.

vitsalis committed:
[NODE_BOT] Use versions

dionyziz commented on bots/node/package.json:
```diff
@@ -0,0 +1,23 @@
+    "body-parser": "^1.6.0",
+    "express": "^4.8.0",
+    "multer": "^0.1.3",
+    "mocha": "^1.4.14",
```
Test requirements should be in `devDependencies`.

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,349 @@
+            .expect( function( res ) {
+                var expData = '{"botname":"sample_botname","version":"0.1.0","username":"sample_username"}';
+
+                assert.strictEqual( expData, res.text );
```
Your assertions should not check for exact serialized equality, but that the JSON objects are deeply equal. This is the bot contract.

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,349 @@
+            .expect( function( res ) {
+                var expData = '{}';
+
+                assert.strictEqual( expData, res.text );
```
Similar.

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,349 @@
+    function requestAndCheckResponse( map, validDirections, validActions, myid, numCreatures, done ) {
+        agent
+            .post( '/round' )
+            .send( 'map=' + encodeURIComponent( JSON.stringify( map ) ) +'&myid=1&W=10&H=10&round=1&gameid=1' )
```
Use dict.

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,349 @@
+            .post( '/round' )
+            .send( 'map=' + encodeURIComponent( JSON.stringify( map ) ) +'&myid=1&W=10&H=10&round=1&gameid=1' )
+            .set( 'Accept', 'application/json' )
+            .expect( 'Content-Type', /json/ )
```
DRY

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,349 @@
+    it( 'should respond with a 400 status code when invalid json is sent', function( done ) {
+        agent
+            .post( '/round' )
+            .send( 'map={fadfas&myid=1&W=10&round=1&gameid=1' )
```
Dict.

vitsalis committed:
[NODE_BOT] Create function tryFn to catch errors

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,349 @@
+                userid: 3,
+                x: 2,
+                y: 1,
+                hp: 100
```
Have some sane defaults, e.g. `hp` is often `100` unless you overwrite it. DRY.

vitsalis committed:
[NODE_BOT] Move test requirements to devDependencies

vitsalis committed:
[NODE_BOT] Send objects instead of strings

vitsalis committed:
[NODE_BOT] Remove useless function and fix a failing test

petrosagg commented on bots/node/lib/bot.js:
```diff
@@ -0,0 +1,121 @@
+            fn( req, res );
+        }
+        catch ( e ) {
+            res.status( 400 ).end();
```
This should be 500

dionyziz commented on bots/node/test/bot.js:
```diff
@@ -0,0 +1,372 @@
+        agent
+            .get( '/v1/bot' )
+            .expect( function( res ) {
+                var expData = {"botname":"sample_botname","version":"0.1.0","username":"sample_username"};
```
nit: style

vitsalis committed:
[NODE_BOT] Fix style

vitsalis committed:
[NODE_BOT] 500 status code on server error
